---
output: pdf_document
---

<!-- Created by Daniel Beck on 7/9/2015 -->
<!-- Modified: -->
<!--      7/10/2015 Continued development -->
<!--      8/10/2015 Modified for automatic report generation -->

```{r, echo=FALSE, message=FALSE, warning=FALSE}
source("dataNames.R")
source("customFunctions.R")
library(pander)
library(bsgenomePackageName, character.only=T, lib.loc=genomeDirectory)
# this file holds information on specific report variables. it is generated by generateReports.R
load("reportValues.Rdata")

# select the desired comparison
an<-which(comparisonNames==analysisName)

# list position for these p-values
pvc<-which(pValues==reportPvalue)
MTCpvc<-which(pValues==MTCreportPvalue)

# load results
load(paste(resultsDirectory, comparisonNames[an], "/methLists.RData", sep=""))
load(paste(resultsDirectory, comparisonNames[an], "/methResults.RData", sep=""))

```


# MeDIP analysis summary for the `r projectName` project (`r comparisonNames[an]`).
Summary generated `r format(Sys.Date(), format="%B %d %Y")`

## Sample comparison
The samples `r paste(seqFiles$sampleName[comparison[[an]]$mset1], collapse=", ")` were compared with the samples `r paste(seqFiles$sampleName[comparison[[an]]$mset2], collapse=", ")`. The samples were analyzed in `r ifelse(comparison[[an]]$pairs[1], 'pairs', 'groups')`. The *`r species`* reference genome was provided by `r provider` and was downloaded from `r genomeSourceFiles`.

```{r, echo=FALSE, results="asis", message=FALSE, warning=FALSE}
if (mapReads){
     pandoc.table(mapExtract(), split.tables=Inf, style="rmarkdown", caption='This table shows the number of reads present for each sample and the overall alignment rate calculated by bowtie2.')
}
```


## Differential coverage analysis
The differential coverage analysis was performed using the MEDIPS R package. The analysis used a windows size of `r ws`bp. DMRs were defined using a `r diff.method` p-value threshold of `r reportPvalue`. DMR edges were extended until there were no genomic windows within `r adjDist`bp with a p-value less than `r dmrBoundPvalue`. Annotation information was downloaded from `r ifelse(annotationType=="biomart", 'the Ensembl database using the biomaRt R package', annotationSource)`.

```{r, echo=FALSE, results="asis", message=FALSE, warning=FALSE}
pandoc.table(as.data.frame(dmrNumberTable), split.tables=110, style="rmarkdown", caption="This table shows the number of DMRs found using different edgeR p-value cutoff thresholds.")

```

************

```{r, echo=FALSE, results="asis", message=FALSE, warning=FALSE}

nWin<-table(methList[[pvc]]$numSigWin)
nWinTable<-as.data.frame(matrix(ncol=length(nWin), nrow=2, c(names(nWin), nWin), byrow=T))
row.names(nWinTable)<-c("Number of significant windows", "Number of DMR")
colnames(nWinTable)<-rep(" ", ncol(nWinTable))
pandoc.table(nWinTable, split.tables=110, style="rmarkdown", caption=paste("This table shows the number of DMR with each specific number of significant windows at a p-value threshold of ", reportPvalue, sep=""))
             
```

************

## Chromosome plot

```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.height=8}
clusterObject<-slidingWindowCluster(formatForCluster(methList2p[[pvc]]), windowLength=2000000, incrementLength=50000)
colnames(clusterObject)<-c("chr", "start", "stop", "minP")
chrLengths<-seqlengths(eval(parse(text=referenceName)))
plotChromosomes(siteTable=methList2p[[pvc]], chrLengths=chrLengths, ymar=5, clusters=clusterObject)
```

**Figure 1.** This figure shows the DMR locations on the individual chromosomes or contigs. Only DMR containing at least two significant windows at a p-value threshold of `r format(reportPvalue,scientific=T)` are shown here.

************

## CpG density histogram

```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.height=5}
par(mar=c(5, 5, 4, 2) + 0.1)
plotCpGdensity(methList2p[[pvc]], main="", xlab="Number of CpG sites per 100bp", ylab="Number of differential DNA \n methylation regions")
```

**Figure 2.** This figure is a histogram of the number of DMRs at different CpG densities. Only DMR containing at least two significant windows at a p-value threshold of `r format(reportPvalue,scientific=T)` are shown here.

************

## DMR length histogram

```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.height=5}
plotDMRlength(methList2p[[pvc]])
```

**Figure 3.** This figure is a histogram of the DMR lengths. Only DMR containing at least two significant windows at a p-value threshold of `r format(reportPvalue,scientific=T)` are shown here.







