---
title: "MeDIP analysis report"
output: pdf_document
---

<!-- Created by Daniel Beck on 7/9/2015 -->
<!-- Modified: -->
<!--      7/10/2015 Continued development -->



```{r, echo=FALSE, message=FALSE, warning=FALSE}
source("dataNames.R")
source("customFunctions.R")
library(pander)
library(bsgenomePackageName, character.only=T, lib.loc=genomeDirectory)


# select the desired comparison
analysis<-"all"
an<-which(comparisonNames==analysis)

# select P-value to use
selectedPvalue<-1e-4
MTCselectedPvalue<-0.05
# list position for these p-values
pvc<-which(pValues==selectedPvalue)
MTCpvc<-which(pValues==MTCselectedPvalue)

# load results
load(paste(resultsDirectory, analysis, "/results.RData", sep=""))

# function to extract mapping information
mapExtract<-function(){
     prepareData<-readLines("prepareData.Rout")
     sn<-c(seqFiles$sampleName[comparison[[an]]$mset1],seqFiles$sampleName[comparison[[an]]$mset2])
     pct<-sapply(strsplit(prepareData[sapply(sn, function(i) grep(pattern=paste("Running Bowtie2 on ", i, '\"', sep=""), x=prepareData))+15], split=" "), function(i) i[1])
     rn<-sapply(strsplit(prepareData[sapply(sn, function(i) grep(pattern=paste("Running Bowtie2 on ", i, '\"', sep=""), x=prepareData))+1], split=" "), function(i) i[1])
     outTable<-as.data.frame(rbind(rn, pct))
     colnames(outTable)<-sn
     rownames(outTable)<-c("read number", "overall alignment rate")
     return(outTable)
}

```

This report summarizes the results from the `r projectName` project. It was generated on `r format(Sys.Date(), format="%B %d %Y")`.

# Sample comparison
The samples `r paste(seqFiles$sampleName[comparison[[an]]$mset1], collapse=", ")` were compared with the samples `r paste(seqFiles$sampleName[comparison[[an]]$mset2], collapse=", ")`. The samples were analyzed in `r ifelse(comparison[[an]]$pairs[1], 'pairs', 'groups')`.

# Raw data cleaning and mapping
The raw data was apparently cleaned to remove low-quality reads by the sequencing lab. It is not clear how this was done. The following table shows the association between the samples and the original data files.

```{r, echo=FALSE, results="asis", message=FALSE, warning=FALSE}
     sn<-c(seqFiles$sampleName[comparison[[an]]$mset1],seqFiles$sampleName[comparison[[an]]$mset2])
     fnTable<-as.data.frame(seqFiles[match(sn, seqFiles$sampleName),])
     row.names(fnTable)<-fnTable$sampleName
     fnTable<-fnTable[,-1]
     colnames(fnTable)<-c("first end filename", "second end filename")
     pandoc.table(fnTable, split.tables=110, style="rmarkdown", caption="This table shows the association between the samples and the raw data files.")
```

A quality report for each raw data file was generated using FastQC and is available in the folder `r dataDirectory`. The raw data consists of 50bp paired end reads. The reads were mapped to the reference genome using Bowtie2.

```{r, echo=FALSE, results="asis", message=FALSE, warning=FALSE}
if (mapReads){
     pandoc.table(mapExtract(), split.tables=Inf, style="rmarkdown", caption='This table shows the number of reads present for each sample and the overall alignment rate calculated by bowtie2.')
}
```

The *`r species`* reference genome was provided by `r provider` and was downloaded from `r genomeSourceFiles`. The mapped reads were converted to the sorted BAM file format using Samtools.

******

# Differential coverage analysis
The differential coverage analysis was performed using the MEDIPS R package. The analysis used a windows size of `r ws`bp. The `r diff.method` method was used to calculate p-values. The `r p.adj` multiple testing adjustment was also calculated. Annotation information was downloaded from `r ifelse(annotationType=="biomart", 'the Ensembl database using the biomaRt R package', annotationSource)`.

```{r, echo=FALSE, results="asis", message=FALSE, warning=FALSE}
# this table can be deleted, it is already calculated by medipAnalysis.
dmrNumberTable<-cbind(
     "p-value"=pValues,
      "singleWindows"=sapply(methList, nrow),
      "multipleWindows"=sapply(methList3p,nrow),
      "MTCsingleWindows"=sapply(MTCmethList, nrow),
      "MTCmultipleWindows"=sapply(MTCmethList3p, nrow)
     )
dmrNumberTable<-as.data.frame(dmrNumberTable)
pandoc.table(dmrNumberTable, split.tables=110, style="rmarkdown", caption="This table shows the number of DMRs found using different p-value cutoff thresholds. The singleWindows and multipleWindows columns use the uncorrected p-value while the MTCsingleWindows and MTCmultipleWindows columns use a multiple testing adjusted p-value.")

```

## Overall p-value distribution
The overall p-value distribution shows a summary of the p-values for each genomic window. Odd patterns in the histogram below may indicate problems with the data or the analysis.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
layout(matrix(c(1,2), nrow=2), heights=c(1,1.5))
par(mar=c(2.1,4.1,1.1,2.1))
a<-hist(methResults$edgeR.p.value, breaks=100, main="", xlab="")
par(mar=c(5.1,4.1,1.1,2.1))
ylim<-sort(a$counts, decreasing=T)[2]
if (is.na(ylim)) ylim<-1
hist(methResults$edgeR.p.value, breaks=100, ylim=c(0, ylim), xlab="edgeR p-value", main="")
```

**Figure 1.** This figure shows the histogram of the edgeR p-values for all genomic windows. The top plot shows the entire histogram while the bottom plot has a truncated y-axis to better show low-frequency p-values.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
layout(matrix(c(1,2), nrow=2), heights=c(1,1.5))
par(mar=c(2.1,4.1,1.1,2.1))
a<-hist(methResults$edgeR.adj.p.value, breaks=100, main="", xlab="")
par(mar=c(5.1,4.1,1.1,2.1))
ylim<-sort(a$counts, decreasing=T)[2]
if (is.na(ylim)) ylim<-1
hist(methResults$edgeR.adj.p.value, breaks=100, ylim=c(0, ylim), xlab=paste(p.adj, " adjusted p-value", sep=""), main="")
```

**Figure 2.** This figure shows the histogram of the `r paste(p.adj, " adjusted p-values", sep="")` for all genomic windows. The top plot shows the entire histogram while the bottom plot has a truncated y-axis to better show low-frequency p-values.

******

## Chromosome plot

```{r, echo=FALSE, message=FALSE, warning=FALSE}
plotChromosomes(siteTable=methList3p[[pvc]], chrLengths=seqlengths(eval(parse(text=referenceName))), markerWidth=4e06, ymar=5)
```

**Figure 3.** This figure shows the DMR locations on the individual chromosomes or contigs. Only multiple window DMRs at a p-value threshold of `r format(selectedPvalue,scientific=T)` are shown here.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
plotChromosomes(siteTable=MTCmethList3p[[MTCpvc]], chrLengths=seqlengths(eval(parse(text=referenceName))), markerWidth=4e06, ymar=5)
```

**Figure 4.** This figure shows the DMR locations on the individual chromosomes or contigs. Only multiple window DMRs at a `r p.adj` adjusted p-value threshold of `r format(MTCselectedPvalue,scientific=T)` are shown here.

******

## CpG density histogram

```{r, echo=FALSE, message=FALSE, warning=FALSE}
plotCpGdensity(methList3p[[pvc]], main="", xlab="Number of CpG sites per 100bp", ylab="Number of differential DNA methylation regions")
```

**Figure 5.** This figure is a histogram of the number of DMRs at different CpG densities. Only multiple window DMRs at a p-value threshold of `r format(selectedPvalue,scientific=T)` are shown here.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
plotCpGdensity(MTCmethList3p[[MTCpvc]], main="", xlab="Number of CpG sites per 100bp", ylab="Number of differential DNA methylation regions")
```

**Figure 6.** This figure is a histogram of the number of DMRs at different CpG densities. Only multiple window DMRs at a `r p.adj` adjusted p-value threshold of `r format(MTCselectedPvalue,scientific=T)` are shown here.