---
output: pdf_document
---

<!-- Created by Daniel Beck on 7/9/2015 -->
<!-- Modified: -->
<!--      7/10/2015 Continued development -->
<!--      8/7/2015 Modified for automatic report generation -->

```{r, echo=FALSE, message=FALSE, warning=FALSE}
source("dataNames.R")
source("customFunctions.R")
library(pander)
library(bsgenomePackageName, character.only=T, lib.loc=genomeDirectory)
# this file holds information on specific report variables. it is generated by generateReports.R
load("reportValues.Rdata")

# select the desired comparison
an<-which(comparisonNames==analysisName)

# list position for these p-values
pvc<-which(pValues==reportPvalue)
MTCpvc<-which(pValues==MTCreportPvalue)

# load results
load(paste(resultsDirectory, comparisonNames[an], "/methLists.RData", sep=""))
load(paste(resultsDirectory, comparisonNames[an], "/methResults.RData", sep=""))


```

# MeDIP analysis summary for the `r projectName` project (`r comparisonNames[an]`).
Report generated `r format(Sys.Date(), format="%B %d %Y")`

## Sample comparison
The samples `r paste(seqFiles$sampleName[comparison[[an]]$mset1], collapse=", ")` were compared with the samples `r paste(seqFiles$sampleName[comparison[[an]]$mset2], collapse=", ")`. The samples were analyzed in `r ifelse(comparison[[an]]$pairs[1], 'pairs', 'groups')`.

## Raw data cleaning and mapping
The raw data was apparently cleaned to remove low-quality reads by the sequencing lab. It is not clear how this was done. The following table shows the association between the samples and the original data files.

```{r, echo=FALSE, results="asis", message=FALSE, warning=FALSE}
     sn<-c(seqFiles$sampleName[comparison[[an]]$mset1],seqFiles$sampleName[comparison[[an]]$mset2])
     fnTable<-as.data.frame(seqFiles[match(sn, seqFiles$sampleName),])
     row.names(fnTable)<-fnTable$sampleName
     fnTable<-fnTable[,-1]
     colnames(fnTable)<-c("first end filename", "second end filename", "sample type")
     pandoc.table(fnTable, split.tables=110, style="rmarkdown", caption="This table shows the association between the samples and the raw data files.")
```

A quality report for each raw data file was generated using FastQC and is available in the folder `r dataDirectory`. The raw data consists of 50bp paired end reads. The reads were mapped to the reference genome using Bowtie2.

```{r, echo=FALSE, results="asis", message=FALSE, warning=FALSE}
if (mapReads){
     pandoc.table(mapExtract(), split.tables=Inf, style="rmarkdown", caption='This table shows the number of reads present for each sample and the overall alignment rate calculated by bowtie2.')
}
```

The *`r species`* reference genome was provided by `r provider` and was downloaded from `r genomeSourceFiles`. The mapped reads were converted to the sorted BAM file format using Samtools.

******

## Differential coverage analysis
The differential coverage analysis was performed using the MEDIPS R package. The analysis used a windows size of `r ws`bp. The `r diff.method` method was used to calculate p-values. The `r p.adj` multiple testing adjustment was also calculated. Annotation information was downloaded from `r ifelse(annotationType=="biomart", 'the Ensembl database using the biomaRt R package', annotationSource)`.

```{r, echo=FALSE, results="asis", message=FALSE, warning=FALSE}

pandoc.table(as.data.frame(dmrNumberTable)[,1:4], split.tables=110, style="rmarkdown", caption="This table shows the number of DMRs found using different p-value cutoff thresholds. The allWindow column shows all DMRs. The twoWindow column shows the number of DMRs containing at least two significant windows. The threeWindow column shows the number of DMRs containing at least three significant windows.")

pandoc.table(as.data.frame(dmrNumberTable)[,c(1,5,6,7)], split.tables=110, style="rmarkdown", caption="This table shows the number of DMRs found using different multiple testing corrected p-value cutoff thresholds. The MTCallWindow column shows all DMRs. The MTCtwoWindow column shows the number of DMRs containing at least two significant windows. The MTCthreeWindow column shows the number of DMRs containing at least three significant windows.")

```

### Overall p-value distribution
The overall p-value distribution shows a summary of the p-values for each genomic window. Odd patterns in the histogram below may indicate problems with the data or the analysis.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
layout(matrix(c(1,2), nrow=2), heights=c(1,1.5))
par(mar=c(2.1,4.1,1.1,2.1))
a<-hist(methResults$edgeR.p.value, breaks=100, main="", xlab="")
par(mar=c(5.1,4.1,1.1,2.1))
ylim<-sort(a$counts, decreasing=T)[2]
if (is.na(ylim)) ylim<-1
hist(methResults$edgeR.p.value, breaks=100, ylim=c(0, ylim), xlab="edgeR p-value", main="")
```

**Figure 1.** This figure shows the histogram of the edgeR p-values for all genomic windows. The top plot shows the entire histogram while the bottom plot has a truncated y-axis to better show low-frequency p-values.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
layout(matrix(c(1,2), nrow=2), heights=c(1,1.5))
par(mar=c(2.1,4.1,1.1,2.1))
a<-hist(methResults$edgeR.adj.p.value, breaks=100, main="", xlab="")
par(mar=c(5.1,4.1,1.1,2.1))
ylim<-sort(a$counts, decreasing=T)[2]
if (is.na(ylim)) ylim<-1
hist(methResults$edgeR.adj.p.value, breaks=100, ylim=c(0, ylim), xlab=paste(p.adj, " adjusted p-value", sep=""), main="")
```

**Figure 2.** This figure shows the histogram of the `r paste(p.adj, " adjusted p-values", sep="")` for all genomic windows. The top plot shows the entire histogram while the bottom plot has a truncated y-axis to better show low-frequency p-values.

******

### Chromosome plot

```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.height=8}
chrLengths<-seqlengths(eval(parse(text=referenceName)))
plotChromosomes(siteTable=methList[[pvc]], chrLengths=chrLengths, ymar=5)
```

**Figure 3.** This figure shows the DMR locations on the individual chromosomes or contigs. All DMRs at a p-value threshold of `r format(reportPvalue,scientific=T)` are shown here.

```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.height=8}
chrLengths<-seqlengths(eval(parse(text=referenceName)))
plotChromosomes(siteTable=methList2p[[pvc]], chrLengths=chrLengths, ymar=5)
```

**Figure 4.** This figure shows the DMR locations on the individual chromosomes or contigs. Only DMRs containing at least two significant windows at a p-value threshold of `r format(reportPvalue,scientific=T)` are shown here.

```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.height=8}
chrLengths<-seqlengths(eval(parse(text=referenceName)))
plotChromosomes(siteTable=methList3p[[pvc]], chrLengths=chrLengths, ymar=5)
```

**Figure 5.** This figure shows the DMR locations on the individual chromosomes or contigs. Only DMRs containing at least three significant windows at a p-value threshold of `r format(reportPvalue,scientific=T)` are shown here.

******

### CpG density histogram

```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.height=5}
par(mar=c(5, 5, 4, 2) + 0.1)
plotCpGdensity(methList[[pvc]], main="", xlab="Number of CpG sites per 100bp", ylab="Number of differential DNA \n methylation regions")
```

**Figure 6.** This figure is a histogram of the number of DMRs at different CpG densities. All DMRs at a p-value threshold of `r format(reportPvalue,scientific=T)` are shown here.

```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.height=5}
par(mar=c(5, 5, 4, 2) + 0.1)
plotCpGdensity(methList2p[[pvc]], main="", xlab="Number of CpG sites per 100bp", ylab="Number of differential DNA \n methylation regions")
```

**Figure 7.** This figure is a histogram of the number of DMRs at different CpG densities. Only DMRs containing at least two significant windows at a p-value threshold of `r format(reportPvalue,scientific=T)` are shown here.


```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.height=5}
par(mar=c(5, 5, 4, 2) + 0.1)
plotCpGdensity(methList3p[[pvc]], main="", xlab="Number of CpG sites per 100bp", ylab="Number of differential DNA \n methylation regions")
```

**Figure 8.** This figure is a histogram of the number of DMRs at different CpG densities. Only DMRs containing at least three significant windows at a p-value threshold of `r format(reportPvalue,scientific=T)` are shown here.

### DMR characteristics


```{r, echo=FALSE, results="asis", message=FALSE, warning=FALSE}

nWin<-table(methList[[pvc]]$numSigWin)
nWinTable<-as.data.frame(matrix(ncol=length(nWin), nrow=2, c(names(nWin), nWin), byrow=T))
row.names(nWinTable)<-c("Number of significant windows", "Number of DMR")
colnames(nWinTable)<-rep(" ", ncol(nWinTable))
pandoc.table(nWinTable, split.tables=110, style="rmarkdown", caption=paste("This table shows the number of DMR with each specific number of significant windows at a p-value threshold of ", reportPvalue, sep=""))
             
```


```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.height=5}
hist(methList[[pvc]]$length, main="", xlab="DMR length (bp)")
```

**Figure 9.** This figure is a histogram of the DMR lengths. All DMRs at a p-value threshold of `r format(reportPvalue,scientific=T)` are shown here.

```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.height=6}
plot(methList[[pvc]]$length, methList[[pvc]]$cpgDensity, xlab="DMR length (bp)", ylab="CpG per 100 bp")
```

**Figure 10.** This figure plots the DMRs by length and CpG density. All DMRs at a p-value threshold of `r format(reportPvalue,scientific=T)` are shown here.


```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.height=6}
plot(methList[[pvc]]$cpgNum, methList[[pvc]]$length, xlab="Number of CpGs", ylab="DMR length (bp)")
```

**Figure 11.** This figure shows the DMR length vs number of CpGs. All DMRs at a p-value threshold of `r format(reportPvalue,scientific=T)` are shown here.



